editor.planx.uk/src/ui/NextStepsList.tsx

```diff
@@ -73,7 +76,16 @@ const ArrowButton = styled("span")(({ theme }) => ({
 }));
 
 const LinkStep = (props: ListItemProps) => (
-  <InnerLink href={props.url} target="_blank" rel="noopener">
+  <InnerLink
+    href={props.url}
+    target="_blank"
+    rel="noopener"
+    onClick={() =>
+      props.handleNextStepsLinkClick &&
+      props?.url &&
+      props.handleNextStepsLinkClick({ url: props.url })
+    }
+  >
```

Comment:
I'm not loving what's going on here although I was getting warnings until I added this. I think it suggests I've made a type error somewhere else 🤔

Reply:
Good question. This is happening because the `url` property is optional in  `ListItemProps` despite actually being required in `LinkStep`.

There's a few ways to expressing this in TypeScript - I'd probably recommend trying to reimplement `ListItemProps` as a discriminated union of two types - one where `url` and `handleNextStepsLinkClick` are not there, and one where they're required. 

There's a nice example of this here - https://www.totaltypescript.com/discriminated-unions-are-a-devs-best-friend

Reply:
Another option (which I think would work?) could be just to call the `useAnalyticsTracking()` hook within this component to save passing the function along as props.

No right or wrong answers here necessarily!

Reply:
> Another option (which I think would work?) could be just to call the `useAnalyticsTracking()` hook within this component to save passing the function along as props.
> 
> No right or wrong answers here necessarily!

I had tried to do this but I think I was encountering issues around using the hook in the component. I suspect it might just be a syntax thing which I'm struggling to recall 😅

Reply:
> Good question. This is happening because the `url` property is optional in `ListItemProps` despite actually being required in `LinkStep`.
> 
> There's a few ways to expressing this in TypeScript - I'd probably recommend trying to reimplement `ListItemProps` as a discriminated union of two types - one where `url` and `handleNextStepsLinkClick` are not there, and one where they're required.
> 
> There's a nice example of this here - https://www.totaltypescript.com/discriminated-unions-are-a-devs-best-friend

I don't think I've seen the union of types pattern before but it looks brilliant! I think it definitely solves for frustrations I've previously had 😌